packages/kbn-alerting-types/alerts_types.ts

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+import { TechnicalRuleDataFieldName } from '@kbn/rule-data-utils';
+import { JsonValue } from '@kbn/utility-types';
+
+export interface MetaAlertFields {
+  _id: string;
+  _index: string;
+}
+
+export interface LegacyField {
+  field: string;
+  value: Array<string | number>;
+}
+
+export interface EsQuerySnapshot {
+  request: string[];
+  response: string[];
+}
+
+export type KnownAlertFields = {
+  [Property in TechnicalRuleDataFieldName]?: JsonValue[];
+};
+
+export type UnknownAlertFields = Record<string, string | JsonValue[]>;
```

Comment:
This "unknown" part of the `Alert` type was previously using `unknown[]` to describe value types:

https://github.com/elastic/kibana/blob/b6287708f687d4e3288851052c0c6ae4ade8ce60/packages/kbn-alerting-types/alert_type.ts#L17-L21

I moved to a more accurate representation of the possible values returned by the Elasticsearch query, using `JsonValue` to indicate that values could be strings, numbers, booleans, objects and arrays (the `[]` is in accordance with the fact that the `fields` query always returns arrays of one or more elements for both single and multi-dimensional values).

The `string |` part is to mitigate a long-standing type error that required us to use a lot of type casts and is caused by an intrinsic limitation of the TypeScript type system with loose types that allow additional properties:

```ts
const alert: Alert = {
  _id: 'test-id',
       ^^^^^^^^^ string is not assignable to type unknown[]
  _index: '.alerts',
  'kibana.alert.status': ['active'],
};
```

A more definitive solution would be to change this alert type and the way we unwrap search responses by finding a more typeable way to represent the static meta fields (perhaps a sub-object?).